haystack/agents/base.py

```diff
@@ -0,0 +1,271 @@
+
+
+class Tool:
+    """
+    A tool is a pipeline or node that also has a name and description. When you add a Tool to an Agent, the Agent can
+    invoke the underlying pipeline or node to answer questions. The wording of the description is important for the
+    Agent to decide which tool is most useful for a given question.
+
+    :param name: The name of the tool. The Agent uses this name to refer to the tool in the text the Agent generates.
+        The name should consist of only one word/token and should be good description of what the tool can do,
+        for example "Calculator" or "Search".
+    :param pipeline_or_node: The pipeline or node to run when this tool is invoked by an Agent.
+    :param description: A description of what the tool is useful for. An Agent can use this description for the decision
+        when to use which tool. An exemplary description of a tool for calculations is "useful for when you need to
+        answer questions about math".
+    """
+
+    def __init__(
+        self,
+        name: str,
+        pipeline_or_node: Union[
+            BaseComponent,
+            Pipeline,
+            ExtractiveQAPipeline,
+            DocumentSearchPipeline,
+            GenerativeQAPipeline,
+            SearchSummarizationPipeline,
+            FAQPipeline,
+            TranslationWrapperPipeline,
+            RetrieverQuestionGenerationPipeline,
+        ],
+        description: str,
+    ):
+        self.name = name
+        self.pipeline_or_node = pipeline_or_node
+        self.description = description
+
+
+class Agent:
+    """
+    An Agent answers queries by choosing between different tools, which are pipelines or nodes. It uses a large
+    language model (LLM) to generate a thought based on the query, choose a tool, and generate the input for the
+    tool. Based on the result returned by the tool the Agent can either stop if it knows the answer now or repeat the
+    process of 1) generate thought, 2) choose tool, 3) generate input.
+
+    Agents are useful for questions containing multiple subquestions that can be answered step-by-step (Multihop QA)
+    using multiple pipelines and nodes as tools.
+
+    :param prompt_node: The PromptNode that the Agent uses to decide which tool to use and what input to provide to it in each iteration.
+    :param tools: A list of Tools that the Agent can choose to run. If no tools are provided, they need to be added with add_tool() before you can use the Agent.
+    :param max_iterations: How many times the agent chooses another tool default 10. If set to 1, only one tool is chosen and there are no further iterations afterward.
+    """
+
+    def __init__(self, prompt_node: PromptNode, tools: Optional[List[Tool]] = None, max_iterations: int = 10):
+        self.prompt_node = prompt_node
+        if tools is not None:
+            self.tools = {tool.name: tool for tool in tools}
+            self.prompt_text = self._generate_prompt_text()
```

Comment:
Thank you so much for the initial feedback. It's all addressed now, except for adding `prompt_text` as an init parameter of the `Agent`. This one needs more discussion.

Reply:
I changed the init method so that users can pass a registered PromptTemplate in the Agent's init method. It just needs to be registered with the PromptNode beforehand, which should be no problem.